### DISCUSSION TITLE
What is the best practice to update atoms when another one is changed?

### DISCUSSION THREAD

**Discussion by michaelschufi** [pmndrs/jotai#2345] [Q&A]
Hi everyone

I'm falling in love with Jotai and its simplicity. Though, I've got one question regarding side effects.

### Use Case & Question
Say we have two checkboxes used to filter a list (`<FilterOne />` and `<FilterTwo />`, with the `filterOneAtom` & `filterTwoAtom`). Those filters are exclusive, which means if `filterOneAtom === true`, `filterTwoAtom` should be `false`.

What's the best way to implement this behaviour using Jotai?

### Implementation Ideas
There are 3 variants that come to mind, regarding where this should be handled:
- Inside the component that updates the atom
- In a side effect handler that gets automatically called whenever an atom updates
- Inside the atom that gets updated (atom updates other atom)

#### Variant 1
The disadvantage here is that, the components updating `filterOneAtom` need to know about all the dependencies the atom has. I would very much like to prevent that.

#### Variant 2
The components updating `filterOneAtom` still have to mount the side effect in their render function. This means it could be forgotten. And in more complex cases it's not obvious how the data flows.

#### Variant 3
This would make the most sense about "where". The atoms are declarative, and so should be the dependencies.

### How to implement variant 3?
Currently I'm using the following pattern

```ts
const filterOneBaseAtom = atom(false);
export const filterOneAtom = atom(
  (get) => get(filterOneBaseAtom),
  (_, set, value: boolean) => {
    set(filterOneBaseAtom, value);
    set(filterTwoBaseAtom, false);
  },
);

const filterTwoBaseAtom = atom(false);
export const filterTwoAtom = atom(
  (get) => get(filterTwoBaseAtom),
  (_, set, value: boolean) => {
    set(filterTwoBaseAtom, value);
    set(filterOneBaseAtom, false);
  },
);
```
([Sandbox](https://codesandbox.io/p/devbox/gracious-hertz-6dptfy?file=%2Fapp%2Ffilters.tsx%3A8%2C42&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clriz56km00072a6cexo8137q%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clriz56km00022a6cy4ltm48r%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clriz56km00042a6cwamlr17x%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clriz56km00062a6cjufwokwr%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clriz56km00022a6cy4ltm48r%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clriz56km00012a6cb3oxsqpn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clrj3komh00022a6cl4p5iecs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A8%252C%2522startColumn%2522%253A42%252C%2522endLineNumber%2522%253A8%252C%2522endColumn%2522%253A42%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Ffilters.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clriz56km00022a6cy4ltm48r%2522%252C%2522activeTabId%2522%253A%2522clrj3komh00022a6cl4p5iecs%2522%257D%252C%2522clriz56km00062a6cjufwokwr%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clriz56km00052a6cpd6epcn4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clriz56km00062a6cjufwokwr%2522%252C%2522activeTabId%2522%253A%2522clriz56km00052a6cpd6epcn4%2522%257D%252C%2522clriz56km00042a6cwamlr17x%2522%253A%257B%2522id%2522%253A%2522clriz56km00042a6cwamlr17x%2522%252C%2522activeTabId%2522%253A%2522clriz56km00032a6ct4aw1fpz%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clriz56km00032a6ct4aw1fpz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522install%2522%252C%2522id%2522%253A%2522clrj7h77d019v2a6c9hrpeqlk%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D))

- Is this ok? 
- I would love to not have to have a base atom if possible. I.e. have a way to override the write and read logic directly in a single atom. Is this possible?
- If I had 5 atoms that all trigger an update (reset) to another, I have to write this logic in all the 5 atoms. Which hurts readability.

I've seen #2136, #1508 and #2003, which, tbh, confused me. And #2136 seems to be partially overkill for what I need to do.

Anyway, this ended in a `jotai-effect` package. So my understanding is I could have an `atomEffect` like

```ts
const onFiltersUpdateAtomEffect = atomEffect((get, set) => {
  const filter1  = get(filter1Atom);
  const filter2  = get(filter2Atom);
  const filter3  = get(filter3Atom);
  const filter4  = get(filter4Atom);

  if (filter1 || filter2 || filter3 || filter4) {
    set(otherAtom, false);
  }
})
```

which I could somehow link to the atoms that need it.  But do I need to have two atoms per atom too? Since whenever I add a `write` function as parameter, I cannot write to the atom itself any more - is this true?

I hope it's somewhat understandable what I would like to achieve.

**Comment by dai-shi** [comment]
> Those filters are exclusive

The best practice is to derive it:
```js
const filterTwoAtom = atom((get) => !get(filterOneAtom));
```
If you need to update from the second one, do so:
```js
const filterTwoAtom = atom(
  (get) => !get(filterOneAtom),
  (get, set, newValue) => set(filterOneAtom, !newValue),
);
```

If this is not acceptable, variant 3 is also a good practice.

**Reply by michaelschufi** [reply]
Thank you for the quick answer!

I realise now, my example is bad. Since booleans just have two states, you can use derived atoms. 

What if the checkboxes are tri-states ones (e.g. `true/false/indeterminate`) and I want to update the one only if the other was just set to true - not false or indeterminate? A derived atom is not possible anymore, correct?

Would the third variant be the only solution? And would I have to have two atoms per atom (so to speak)?

**Reply by dai-shi** [reply]
I would try to create one atom as the source of truth and derive from it.

**Reply by michaelschufi** [reply]
Like I do in my first code example?

**Reply by dai-shi** [reply]
It's something like this:
```js
const sourceOfTruthAtom = atom({ first: true, second: 'intermediate' });
```
